agent/tests/Feature/IngestTest.php

```diff
@@ -0,0 +1,868 @@
+        new Timer(interval: 10, runAt: 10, scheduledAt: 0, scheduledBy: 'Laravel\NightwatchAgent\Ingest::write'),
+    ]);
+    expect($loop)->toHavePending([
+        new Timer(interval: 3_600, runAt: 3_600, scheduledAt: 0, scheduledBy: 'Laravel\NightwatchAgent\IngestDetailsRepository::scheduleRefreshIn'),
+    ]);
+    expect($ingestDetailsBrowser)->toHaveSent([
+        Request::json('/api/agent-auth'),
+    ]);
+    expect($ingestDetailsBrowser)->toHavePending([]);
+});
+
+it('handles runtime exceptions while procesing the request', function () {
+    $loop = new LoopFake(runForSeconds: 11);
+    $server = new TcpServerFake;
+    $ingestDetailsBrowser = new BrowserFake([
+        Response::jwt(),
+    ]);
+    $ingestBrowser = new BrowserFake([
+        Response::throwWhileProcessing('Whoops!'),
+    ]);
+    $loop->addTimer(0, $server->pendingConnection([['t' => 'request']]));
+
+    [$output, $e] = run(
+        via: 'source',
+        ingestDetailsBrowser: $ingestDetailsBrowser,
+        ingestBrowser: $ingestBrowser,
+        loop: $loop,
+        server: $server,
+    );
+
+    expect($e)->toBeNull($e?->getMessage() ?? '');
+    expect($output)->toMatchLog(<<<'OUTPUT'
+        {date} {info} Authentication successful {duration}
+        {date} {info} Ingest failed {duration}: Whoops!
+        OUTPUT);
+    expect($ingestBrowser)->toHaveSent([
+        Request::ingest([['t' => 'request']]),
+    ]);
+    expect($ingestBrowser)->toHavePending([]);
+    expect($loop)->toHaveRun([
+        new Timer(interval: 0, runAt: 0, scheduledAt: 0, scheduledBy: self::class),
+        new Timer(interval: 10, runAt: 10, scheduledAt: 0, scheduledBy: 'Laravel\NightwatchAgent\Ingest::write'),
+    ]);
+    expect($loop)->toHavePending([
+        new Timer(interval: 3_600, runAt: 3_600, scheduledAt: 0, scheduledBy: 'Laravel\NightwatchAgent\IngestDetailsRepository::scheduleRefreshIn'),
+    ]);
+    expect($ingestDetailsBrowser)->toHaveSent([
+        Request::json('/api/agent-auth'),
+    ]);
+    expect($ingestDetailsBrowser)->toHavePending([]);
+});
+
+it('handles missing authentication details', function () {
+    $loop = new LoopFake(runForSeconds: 11);
+    $server = new TcpServerFake;
+    $ingestDetailsBrowser = new BrowserFake([
+        Response::unauthenticated(),
+    ]);
+    $ingestBrowser = new BrowserFake([]);
+    $loop->addTimer(0, $server->pendingConnection([['t' => 'request']]));
+
+    [$output, $e] = run(
+        via: 'source',
+        ingestDetailsBrowser: $ingestDetailsBrowser,
+        ingestBrowser: $ingestBrowser,
+        loop: $loop,
+        server: $server,
+    );
+
+    expect($e)->toBeNull($e?->getMessage() ?? '');
+    expect($output)->toMatchLog(<<<'OUTPUT'
+        {date} {info} Authentication failed {duration}: 401 \[{"message":"Invalid environment token"}\]
+        {date} {info} Ingest failed {duration}: No authentication details
+        OUTPUT);
+    expect($ingestBrowser)->toHaveSent([]);
```

Comment:
[Non-blocking]
I think it would be cool if we could do `expect($ingestBrowser)->toHaveSentNothing();` 🙂

Reply:
Love it. Updated.